Enable MCG in the OSD deployer

ezio-auditore opened the pull request:
MCG will only be enabled if the addon params contain enable-mcg(bool) true
Signed-off-by: Kaustav Majumder kmajumder@redhat.com

ezio-auditore commented on controllers/managedocs_controller.go:
```diff
@@ -567,11 +568,22 @@ func (r *ManagedOCSReconciler) updateStorageClusterFromAddonParamsSecret(sc *ocs
 	if err != nil {
 		return fmt.Errorf("Invalid storage cluster size value: %v", sizeAsString)
 	}
+	if enableMcgAsString != "" {
+		desiredMcgEnable, err := strconv.ParseBool(enableMcgAsString)
+		if err != nil {
+			return fmt.Errorf("Invalid Enable MCG value: %v", enableMcgAsString)
+		}
+		if desiredMcgEnable {
+			r.Log.Info("Enabling MCG")
+			sc.Spec.MultiCloudGateway.ReconcileStrategy = "manage"
+		}
+	}
```
Moving this section to end of the function causes 18 unit tests to fail since none of them are  expecting the new parameter
`enable-mcg`. I feel we can do two things
1. fix all the failing tests to include a dummy value for the new parameter 
2. Move this to a new reconcile function and make it execute just after` updateStorageClusterFromAddonParamsSecret()` is called, so the new addon won't disrupt existing tests.

nb-ohad replied:
I believe it happens not because of order, but because the parameter is a non optional one (as opposed to the original code where you have introduced it as optional)

And yes, when we introduce non-optional parameters we need to make changes to both unit-tests and makefile (to make sure `make run` is working correctly) because failing on a missing parameter is the desired and expected behavior

nb-ohad commented on controllers/managedocs_controller.go:
```diff
@@ -567,7 +568,9 @@ func (r *ManagedOCSReconciler) updateStorageClusterFromAddonParamsSecret(sc *ocs
 	sizeAsString := string(addonParams[storageClassSizeKey])
-	r.Log.Info("Requested add-on settings", storageClassSizeKey, sizeAsString)
+	// Setting hardcoded value here to force no MCG deployment
+	enableMCGAsString := "false"
```
```suggestion
	// Setting hardcoded value here to force no MCG deployment
	// enableMCGAsString := string(addonParams[enableMCGKey])
	enableMCGAsString := "false"
```

ezio-auditore replied:
Applied recommended changes

nb-ohad commented on controllers/managedocs_controller_test.go:
```diff
@@ -389,6 +389,25 @@ var _ = Describe("ManagedOCS controller", func() {
 				Expect(k8sClient.Delete(ctx, secret)).Should(Succeed())
 			})
 		})
+		When("there is no enable-mcg field in the add-on parameters secret", func() {
+			It("should not create a reconciled resources", func() {
+				// Create  add-on parameters secret with size field only
+				secret := addonParamsSecretTemplate.DeepCopy()
+				secret.Data["size"] = []byte("1")
+				Expect(k8sClient.Create(ctx, secret)).Should(Succeed())
+
+				// Ensure, over a period of time, that the resources are not created
+				resList := []runtime.Object{
+					scTemplate.DeepCopy(),
+					promTemplate.DeepCopy(),
+					amTemplate.DeepCopy(),
+				}
+				utils.EnsureNoResources(k8sClient, ctx, resList, timeout, interval)
+
+				// Remove the secret for future cases
+				Expect(k8sClient.Delete(ctx, secret)).Should(Succeed())
+			})
+		})
```
After hardcoding the `enable-mcg` param to false, this test should fail.

ezio-auditore replied:
modified the test in the next patch

nb-ohad commented on controllers/managedocs_controller_test.go:
```diff
@@ -408,6 +427,26 @@ var _ = Describe("ManagedOCS controller", func() {
 				Expect(k8sClient.Delete(ctx, secret)).Should(Succeed())
 			})
 		})
+		When("there is an invalid enable-mcg value in the add-on parameters secret", func() {
+			It("should not create reconciled resources", func() {
+				// Create a invalid enable-mcg parameter value
+				secret := addonParamsSecretTemplate.DeepCopy()
+				secret.Data["size"] = []byte("1")
+				secret.Data["enable-mcg"] = []byte("NO")
+				Expect(k8sClient.Create(ctx, secret)).Should(Succeed())
+
+				// Ensure, over a period of time, that the resources are not created
+				resList := []runtime.Object{
+					scTemplate.DeepCopy(),
+					promTemplate.DeepCopy(),
+					amTemplate.DeepCopy(),
+				}
+				utils.EnsureNoResources(k8sClient, ctx, resList, timeout, interval)
+
+				// Remove the secret for future cases
+				Expect(k8sClient.Delete(ctx, secret)).Should(Succeed())
+			})
+		})
```
Same comment as the last one

ezio-auditore replied:
modified the test in the next patch

nb-ohad reviewed (changes requested):
A couple of general comments: 
- Missing a test case for validation of enabling of mcg (verifying changes to `storagecluster`) 
- Missing a test case for the prevention of disabling MCG after it has been enabled

renan-campos commented:
LGTM

ezio-auditore commented:
> A couple of general comments:
> 
> * Missing a test case for validation of enabling of mcg (verifying changes to `storagecluster`)
added in next patch
> * Missing a test case for the prevention of disabling MCG after it has been enabled
Unable to add any relevant test since MCG cannot be enabled at all since the value is hardcoded. The new tests I have added confirms that. Once we remove the hardcoded `enableMCGAsString := "false"` writing this test should work.

nb-ohad commented:
@ezio-auditore Can we write the test, verify it, then hide it behind an hardcoded `if` or a comment?

ezio-auditore commented:
> @ezio-auditore Can we write the test, verify it, then hide it behind an hardcoded `if` or a comment?

I have added `Skip` to this particular test till MCG is properly integrated into Deployer.

nb-ohad commented on controllers/managedocs_controller_test.go:
```diff
@@ -434,11 +434,96 @@ var _ = Describe("ManagedOCS controller", func() {
 			})
 		})
+		When("there is no enable-mcg field in the add-on parameters secret", func() {
+			It("should not create reconciled resources", func() {
```
As you only check for storagecluster, it is best to explicitly specify it:
```suggestion
			It("should not create storagecluster", func() {
```

nb-ohad commented on controllers/managedocs_controller_test.go:
```diff
@@ -434,11 +434,96 @@ var _ = Describe("ManagedOCS controller", func() {
+			})
+		})
+		When("there is an invalid enable-mcg value in the add-on parameters secret", func() {
+			It("should not create reconciled resources", func() {
```
Same as last comment

nb-ohad commented on controllers/managedocs_controller_test.go:
```diff
@@ -434,11 +434,96 @@ var _ = Describe("ManagedOCS controller", func() {
+			})
+		})
+		When("there is a valid enable-mcg value in the addon-on parameters secret which is false", func() {
+			It("should not change the storage cluster spec", func() {
```
Again this is not specific enough
```suggestion
			It("Should change storagecluster's MCG reconcile strategy to manage ", func() {
```

nb-ohad commented on controllers/managedocs_controller_test.go:
```diff
@@ -434,11 +434,96 @@ var _ = Describe("ManagedOCS controller", func() {
+			})
+		})
+		When("there is a valid enable-mcg value in the addon-on parameters secret which is true", func() {
+			It("should change the storage cluster spec to enable MCG", func() {
```
See last comment

nb-ohad commented on controllers/managedocs_controller_test.go:
```diff
@@ -434,11 +434,96 @@ var _ = Describe("ManagedOCS controller", func() {
+			})
+		})
+		When("MCG is already enabled and enable-mcg value in addon-on parameter secret is false", func() {
+			It("should not disable MCG in storage cluster spec", func() {
```
see last comment

nb-ohad commented on controllers/managedocs_controller_test.go:
```diff
@@ -434,11 +434,96 @@ var _ = Describe("ManagedOCS controller", func() {
+						return false
+					}
+					return sc.Spec.MultiCloudGateway.ReconcileStrategy == "manage" ||
+						sc.Spec.MultiCloudGateway.ReconcileStrategy == ""
```
`""` is not an acceptable value in this case, we need to look specifically for "manage"

nb-ohad commented:
/lgtm
